geo: enforce valid SRID for Geography and Geometry types

otan opened the pull request:
Enforce that SRIDs are valid for Geometry/Geometry types, as well as
that they are lat/lng systems for Geography types.

This involved privatising the SpatialObject field for Geometry/Geography
types so that all operations must go through `NewGeo*`. There is an
unsafe option when we trust the object (when it is from the database).

Release note: None

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 15 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50065#-:-M9ck5-p285uQWDXts_7:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geo.go, line 178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50065#-M9cgLAy6PmR5I5Djolg:-M9cgLAy6PmR5I5Djolh:b-mlaiku) ([raw file](https://github.com/cockroachdb/cockroach/blob/8605b97086690c28ace57503940a37f97de62e76/pkg/geo/geo.go#L178)):*
> ```Go
> }
> 
> // ParseGeometryFromEWKBRaw returns a new Geometry from an EWKB, without any SRID checks.
> ```

the "without any SRID checks" is not quite true, since `NewGeometry` will check the SRID.
Same for `ParseGeographyFromEWKBRaw`.
___
*[pkg/geo/geo.go, line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50065#-M9cfYpP5kF_3YOSA9vQ:-M9cfYpP5kF_3YOSA9vR:b-3sqdax) ([raw file](https://github.com/cockroachdb/cockroach/blob/8605b97086690c28ace57503940a37f97de62e76/pkg/geo/geo.go#L180)):*
> ```Go
> // ParseGeometryFromEWKBRaw returns a new Geometry from an EWKB, without any SRID checks.
> // You should only do this if you trust the EWKB is setup correctly.
> // You must likely want geo.ParseGeometryFromEWKB instead.
> ```

s/must/most/
___
*[pkg/geo/geo.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50065#-M9ch-DJ9hA4wzbRreud:-M9ch-DJ9hA4wzbRreue:b-3sqdax) ([raw file](https://github.com/cockroachdb/cockroach/blob/8605b97086690c28ace57503940a37f97de62e76/pkg/geo/geo.go#L377)):*
> ```Go
> // ParseGeographyFromEWKBRaw returns a new Geography from an EWKB, without any SRID checks.
> // You should only do this if you trust the EWKB is setup correctly.
> // You must likely want ParseGeographyFromEWKB instead.
> ```

s/must/most/


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50065)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geo.go, line 178 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50065#-M9cgLAy6PmR5I5Djolg:-M9d1AuvDb9nHgAXZo4_:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/8605b97086690c28ace57503940a37f97de62e76/pkg/geo/geo.go#L178)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

the "without any SRID checks" is not quite true, since `NewGeometry` will check the SRID.
Same for `ParseGeographyFromEWKBRaw`.
</blockquote></details>

Fixed
___
*[pkg/geo/geo.go, line 180 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50065#-M9cfYpP5kF_3YOSA9vQ:-M9d1FjM9B77O-8czZSL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8605b97086690c28ace57503940a37f97de62e76/pkg/geo/geo.go#L180)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/must/most/
</blockquote></details>

Done.
___
*[pkg/geo/geo.go, line 377 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50065#-M9ch-DJ9hA4wzbRreud:-M9d1GWF23KgA0h-Ki3m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8605b97086690c28ace57503940a37f97de62e76/pkg/geo/geo.go#L377)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

s/must/most/
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

otan committed:
geo: enforce valid SRID for Geography and Geometry types

Enforce that SRIDs are valid for Geometry/Geometry types, as well as
that they are lat/lng systems for Geography types.

This involved privatising the SpatialObject field for Geometry/Geography
types so that all operations must go through `NewGeo*`. There is an
unsafe option when we trust the object (when it is from the database).

Release note: None

otan commented:
bors r=sumeerbhola
